crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -474,44 +498,54 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
                 }
             };
 
-            reverts
-                .entry(block_number)
+            revert_storage
+                .entry(block_and_address)
                 .or_default()
-                .entry(address)
-                .or_default()
-                .1
-                .push(old_storage);
+                .push((old_storage.key.into(), old_storage.value));
+
+            bundle_builder = bundle_builder
+                .state_storage(address, state_storage.entry(address).or_default().clone());
+            bundle_builder = bundle_builder.revert_storage(
+                block_number,
+                address,
+                revert_storage.entry(block_and_address).or_default().clone(),
+            );
         }
 
+        let bundle_state = bundle_builder.build();
+
         if TAKE {
             // iterate over local plain state remove all account and all storages.
-            for (address, (old_account, new_account, storage)) in state.iter() {
+            for (address, bundle_account) in bundle_state.state.iter() {
                 // revert account if needed.
-                if old_account != new_account {
+                if bundle_account.is_info_changed() {
                     let existing_entry = plain_accounts_cursor.seek_exact(*address)?;
-                    if let Some(account) = old_account {
-                        plain_accounts_cursor.upsert(*address, *account)?;
+                    if let Some(account) = &bundle_account.original_info {
+                        plain_accounts_cursor.upsert(*address, into_reth_acc(account.clone()))?;
                     } else if existing_entry.is_some() {
                         plain_accounts_cursor.delete_current()?;
                     }
                 }
 
                 // revert storages
-                for (storage_key, (old_storage_value, _new_storage_value)) in storage {
-                    let storage_entry =
-                        StorageEntry { key: *storage_key, value: *old_storage_value };
+                for (storage_key, storage_slot) in bundle_account.storage.iter() {
+                    let b256_key: B256 = (*storage_key).into();
```

Comment:
```suggestion
                    let storage_key: B256 = (*storage_key).into();
```
overwrite the `storage_key` var, `b256_key` is a less helpful name

Reply:
Updated